File: frontend/src/services/FirebaseService.js
import { database } from '../config/firebase';
import { ref, set, get, update, remove, push, query, orderByChild, equalTo } from 'firebase/database';

// Helper to generate IDs
const generateId = () => push(ref(database)).key;

export const FirebaseService = {
  // Exam operations
  createExam: async (examData) => {
    const examId = generateId();
    const now = new Date().toISOString();
    const newExam = {
      id: examId,
      title: examData.title || 'Untitled Exam',
      description: examData.description || '',
      audio_url: examData.audio_url,
      audio_source_method: examData.audio_source_method,
      loop_audio: examData.loop_audio || false,
      duration_seconds: examData.duration_seconds || 1800,
      published: examData.published || false,
      created_at: now,
      updated_at: now,
      is_demo: examData.is_demo || false,
      question_count: 0,
      submission_count: 0,
    };

    await set(ref(database, `exams/${examId}`), newExam);
    
    // Create 4 sections for the exam
    for (let i = 1; i <= 4; i++) {
      const sectionId = generateId();
      await set(ref(database, `sections/${sectionId}`), {
        id: sectionId,
        exam_id: examId,
        index: i,
        title: `Section ${i}`,
      });
    }
    
    return newExam;
  },

  getExam: async (examId) => {
    const snapshot = await get(ref(database, `exams/${examId}`));
    return snapshot.exists() ? snapshot.val() : null;
  },

  getAllExams: async () => {
    const snapshot = await get(ref(database, 'exams'));
    if (!snapshot.exists()) return [];
    const exams = [];
    snapshot.forEach((child) => {
      exams.push(child.val());
    });
    return exams;
  },

  getPublishedExams: async () => {
    const allExams = await FirebaseService.getAllExams();
    return allExams.filter(exam => exam.published);
  },

  updateExam: async (examId, examData) => {
    const updates = {
      ...examData,
      updated_at: new Date().toISOString(),
    };
    await update(ref(database, `exams/${examId}`), updates);
    return await FirebaseService.getExam(examId);
  },

  deleteExam: async (examId) => {
    // Don't allow deleting demo exams
    const exam = await FirebaseService.getExam(examId);
    if (exam?.is_demo) {
      throw new Error('Demo exams cannot be deleted');
    }

    // Delete exam
    await remove(ref(database, `exams/${examId}`));
    
    // Delete related sections
    const sections = await FirebaseService.getSectionsByExam(examId);
    for (const section of sections) {
      await remove(ref(database, `sections/${section.id}`));
      
      // Delete questions in this section
      const questions = await FirebaseService.getQuestionsBySection(section.id);
      for (const question of questions) {
        await remove(ref(database, `questions/${question.id}`));
      }
    }
    
    return true;
  },

  publishExam: async (examId) => {
    return await FirebaseService.updateExam(examId, { published: true });
  },

  // Section operations
  getSectionsByExam: async (examId) => {
    const snapshot = await get(ref(database, 'sections'));
    if (!snapshot.exists()) return [];
    const sections = [];
    snapshot.forEach((child) => {
      const section = child.val();
      if (section.exam_id === examId) {
        sections.push(section);
      }
    });
    return sections.sort((a, b) => a.index - b.index);
  },

  getSection: async (sectionId) => {
    const snapshot = await get(ref(database, `sections/${sectionId}`));
    return snapshot.exists() ? snapshot.val() : null;
  },

  updateSection: async (sectionId, sectionData) => {
    await update(ref(database, `sections/${sectionId}`), sectionData);
    return await FirebaseService.getSection(sectionId);
  },

  // Question operations
  createQuestion: async (questionData) => {
    const section = await FirebaseService.getSection(questionData.section_id);
    if (!section) throw new Error('Section not found');

    // Check if section already has 10 questions
    const sectionQuestions = await FirebaseService.getQuestionsBySection(section.id);
    if (sectionQuestions.length >= 10) {
      throw new Error('Maximum of 10 questions per section');
    }

    const questionId = generateId();
    const newQuestion = {
      id: questionId,
      exam_id: questionData.exam_id,
      section_id: questionData.section_id,
      index: questionData.index || sectionQuestions.length + 1,
      type: questionData.type || 'single_answer',
      payload: questionData.payload || {},
      marks: questionData.marks || 1,
      created_by: questionData.created_by || 'admin',
      is_demo: questionData.is_demo || false,
    };

    await set(ref(database, `questions/${questionId}`), newQuestion);

    // Update question count on exam
    const exam = await FirebaseService.getExam(questionData.exam_id);
    if (exam) {
      await FirebaseService.updateExam(questionData.exam_id, {
        question_count: (exam.question_count || 0) + 1,
      });
    }

    return newQuestion;
  },

  getQuestion: async (questionId) => {
    const snapshot = await get(ref(database, `questions/${questionId}`));
    return snapshot.exists() ? snapshot.val() : null;
  },

  getQuestionsBySection: async (sectionId) => {
    const snapshot = await get(ref(database, 'questions'));
    if (!snapshot.exists()) return [];
    const questions = [];
    snapshot.forEach((child) => {
      const question = child.val();
      if (question.section_id === sectionId) {
        questions.push(question);
      }
    });
    return questions.sort((a, b) => a.index - b.index);
  },

  updateQuestion: async (questionId, questionData) => {
    const question = await FirebaseService.getQuestion(questionId);
    if (question?.is_demo) {
      throw new Error('Demo questions cannot be modified');
    }
    await update(ref(database, `questions/${questionId}`), questionData);
    return await FirebaseService.getQuestion(questionId);
  },

  deleteQuestion: async (questionId) => {
    const question = await FirebaseService.getQuestion(questionId);
    if (question?.is_demo) {
      throw new Error('Demo questions cannot be deleted');
    }

    await remove(ref(database, `questions/${questionId}`));

    // Update question count on exam
    if (question) {
      const exam = await FirebaseService.getExam(question.exam_id);
      if (exam && exam.question_count) {
        await FirebaseService.updateExam(question.exam_id, {
          question_count: exam.question_count - 1,
        });
      }
    }

    return true;
  },

  // Get exam with sections and questions
  getExamWithSectionsAndQuestions: async (examId) => {
    const exam = await FirebaseService.getExam(examId);
    if (!exam) return null;

    const sections = await FirebaseService.getSectionsByExam(examId);
    const sectionsWithQuestions = await Promise.all(
      sections.map(async (section) => {
        const questions = await FirebaseService.getQuestionsBySection(section.id);
        return {
          ...section,
          questions,
        };
      })
    );

    return {
      exam,
      sections: sectionsWithQuestions,
    };
  },

  // Audio management
  uploadAudio: async (examId, audioUrl, method) => {
    return await FirebaseService.updateExam(examId, {
      audio_url: audioUrl,
      audio_source_method: method,
    });
  },

  // Submission operations
  createSubmission: async (submissionData) => {
    const submissionId = generateId();
    const now = new Date().toISOString();
    const newSubmission = {
      id: submissionId,
      exam_id: submissionData.exam_id,
      user_id_or_session: submissionData.user_id_or_session || generateId(),
      started_at: now,
      answers: submissionData.answers || {},
      progress_percent: 0,
      last_playback_time: 0,
    };

    await set(ref(database, `submissions/${submissionId}`), newSubmission);

    // Update submission count on exam
    const exam = await FirebaseService.getExam(submissionData.exam_id);
    if (exam) {
      await FirebaseService.updateExam(submissionData.exam_id, {
        submission_count: (exam.submission_count || 0) + 1,
      });
    }

    return newSubmission;
  },

  updateSubmission: async (submissionId, submissionData) => {
    await update(ref(database, `submissions/${submissionId}`), submissionData);
    const snapshot = await get(ref(database, `submissions/${submissionId}`));
    return snapshot.exists() ? snapshot.val() : null;
  },

  getSubmission: async (submissionId) => {
    const snapshot = await get(ref(database, `submissions/${submissionId}`));
    return snapshot.exists() ? snapshot.val() : null;
  },
};

File: backend/firebase_service.py
"""
Firebase Realtime Database Service
Handles all database operations for exams, questions, sections, and submissions
"""

import firebase_admin
from firebase_admin import db
from typing import List, Optional, Dict, Any
import uuid
from datetime import datetime, timezone
import os
import json

# Initialize Firebase Admin SDK
def init_firebase():
    """Initialize Firebase Admin SDK"""
    try:
        # Check if already initialized
        firebase_admin.get_app()
    except ValueError:
        # Not initialized, so initialize it
        # Use service account key from environment or file
        service_account_key = os.environ.get('FIREBASE_SERVICE_ACCOUNT_KEY')

        try:
            if service_account_key:
                # Parse JSON from environment variable
                cred_dict = json.loads(service_account_key)
                cred = firebase_admin.credentials.Certificate(cred_dict)
            else:
                # Try to load from file - use multiple paths
                cred = None
                possible_paths = [
                    'firebase-key.json',
                    os.path.join(os.path.dirname(__file__), 'firebase-key.json'),
                    '../firebase-key.json',
                    os.path.join(os.path.dirname(__file__), '..', 'firebase-key.json'),
                ]

                for path in possible_paths:
                    if os.path.exists(path):
                        try:
                            cred = firebase_admin.credentials.Certificate(path)
                            print(f"✅ Firebase credentials loaded from: {path}")
                            break
                        except Exception as e:
                            print(f"Failed to load from {path}: {e}")
                            continue

                if cred is None:
                    # Use default credentials (for local development)
                    print("Using Application Default Credentials")
                    cred = firebase_admin.credentials.ApplicationDefault()

            firebase_admin.initialize_app(cred, {
                'databaseURL': os.environ.get('FIREBASE_DATABASE_URL',
                                             'https://ielts-listening-module-default-rtdb.firebaseio.com')
            })
            print("✅ Firebase initialized successfully!")
        except Exception as e:
            print(f"Warning: Firebase initialization failed: {e}")
            print("Firebase operations may not work without proper credentials.")

# Initialize on import
init_firebase()

class FirebaseService:
    """Service for Firebase Realtime Database operations"""

    # ============================================
    # EXAM OPERATIONS
    # ============================================

    @staticmethod
    def create_exam(exam_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new exam"""
        exam_id = str(uuid.uuid4())
        now = datetime.now(timezone.utc).isoformat()

        exam = {
            'id': exam_id,
            'title': exam_data.get('title', 'Untitled Exam'),
            'description': exam_data.get('description', ''),
            'exam_type': exam_data.get('exam_type', 'listening'),
            'audio_url': exam_data.get('audio_url'),
            'audio_source_method': exam_data.get('audio_source_method'),
            'loop_audio': exam_data.get('loop_audio', False),
            'duration_seconds': exam_data.get('duration_seconds', 1800),
            'published': exam_data.get('published', False),
            'created_at': now,
            'updated_at': now,
            'is_demo': exam_data.get('is_demo', False),
            'question_count': 0,
            'submission_count': 0,
            'is_active': exam_data.get('is_active', False),
            'is_visible': exam_data.get('is_visible', True),
        }

        db.reference(f'exams/{exam_id}').set(exam)

        # Create 4 sections for the exam
        for i in range(1, 5):
            section_id = str(uuid.uuid4())
            section = {
                'id': section_id,
                'exam_id': exam_id,
                'index': i,
                'title': f'Section {i}',
            }
            db.reference(f'sections/{section_id}').set(section)

        return exam

    @staticmethod
    def get_exam(exam_id: str) -> Optional[Dict[str, Any]]:
        """Get exam by ID"""
        snapshot = db.reference(f'exams/{exam_id}').get()
        return snapshot

    @staticmethod
    def get_all_exams() -> List[Dict[str, Any]]:
        """Get all exams"""
        snapshot = db.reference('exams').get()
        if not snapshot:
            return []
        return list(snapshot.values()) if isinstance(snapshot, dict) else []

    @staticmethod
    def get_published_exams() -> List[Dict[str, Any]]:
        """Get all published exams"""
        all_exams = FirebaseService.get_all_exams()
        return [exam for exam in all_exams if exam.get('published', False)]

    @staticmethod
    def update_exam(exam_id: str, exam_data: Dict[str, Any]) -> Dict[str, Any]:
        """Update exam"""
        exam_data['updated_at'] = datetime.now(timezone.utc).isoformat()
        db.reference(f'exams/{exam_id}').update(exam_data)
        return FirebaseService.get_exam(exam_id)

    @staticmethod
    def delete_exam(exam_id: str) -> bool:
        """Delete exam and related data"""
        exam = FirebaseService.get_exam(exam_id)
        if exam and exam.get('is_demo'):
            raise ValueError('Demo exams cannot be deleted')

        # Delete exam
        db.reference(f'exams/{exam_id}').delete()

        # Delete sections
        sections = FirebaseService.get_sections_by_exam(exam_id)
        for section in sections:
            db.reference(f'sections/{section["id"]}').delete()

            # Delete questions
            questions = FirebaseService.get_questions_by_section(section['id'])
            for question in questions:
                db.reference(f'questions/{question["id"]}').delete()

        return True

    # ============================================
    # SECTION OPERATIONS
    # ============================================

    @staticmethod
    def get_sections_by_exam(exam_id: str) -> List[Dict[str, Any]]:
        """Get all sections for an exam"""
        snapshot = db.reference('sections').get()
        if not snapshot:
            return []

        sections = [s for s in snapshot.values() if s.get('exam_id') == exam_id]
        return sorted(sections, key=lambda x: x.get('index', 0))

    @staticmethod
    def get_section(section_id: str) -> Optional[Dict[str, Any]]:
        """Get section by ID"""
        return db.reference(f'sections/{section_id}').get()

    @staticmethod
    def update_section(section_id: str, section_data: Dict[str, Any]) -> Dict[str, Any]:
        """Update section"""
        db.reference(f'sections/{section_id}').update(section_data)
        return FirebaseService.get_section(section_id)

    # ============================================
    # QUESTION OPERATIONS
    # ============================================

    @staticmethod
    def create_question(question_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new question"""
        section = FirebaseService.get_section(question_data['section_id'])
        if not section:
            raise ValueError('Section not found')

        # Check max questions per section
        section_questions = FirebaseService.get_questions_by_section(section['id'])
        if len(section_questions) >= 10:
            raise ValueError('Maximum of 10 questions per section')

        question_id = str(uuid.uuid4())
        question = {
            'id': question_id,
            'exam_id': question_data['exam_id'],
            'section_id': question_data['section_id'],
            'index': question_data.get('index', len(section_questions) + 1),
            'type': question_data.get('type', 'single_answer'),
            'payload': question_data.get('payload', {}),
            'marks': question_data.get('marks', 1),
            'created_by': question_data.get('created_by', 'admin'),
            'is_demo': question_data.get('is_demo', False),
        }

        db.reference(f'questions/{question_id}').set(question)

        # Update question count
        exam = FirebaseService.get_exam(question_data['exam_id'])
        if exam:
            FirebaseService.update_exam(question_data['exam_id'], {
                'question_count': exam.get('question_count', 0) + 1
            })

        return question

    @staticmethod
    def get_question(question_id: str) -> Optional[Dict[str, Any]]:
        """Get question by ID"""
        return db.reference(f'questions/{question_id}').get()

    @staticmethod
    def get_questions_by_section(section_id: str) -> List[Dict[str, Any]]:
        """Get all questions for a section"""
        snapshot = db.reference('questions').get()
        if not snapshot:
            return []

        questions = [q for q in snapshot.values() if q.get('section_id') == section_id]
        return sorted(questions, key=lambda x: x.get('index', 0))

    @staticmethod
    def update_question(question_id: str, question_data: Dict[str, Any]) -> Dict[str, Any]:
        """Update question"""
        db.reference(f'questions/{question_id}').update(question_data)
        return FirebaseService.get_question(question_id)

    @staticmethod
    def delete_question(question_id: str) -> bool:
        """Delete question"""
        question = FirebaseService.get_question(question_id)
        if question and question.get('is_demo'):
            raise ValueError('Demo questions cannot be deleted')

        db.reference(f'questions/{question_id}').delete()

        # Update question count
        if question:
            exam = FirebaseService.get_exam(question['exam_id'])
            if exam and exam.get('question_count', 0) > 0:
                FirebaseService.update_exam(question['exam_id'], {
                    'question_count': exam['question_count'] - 1
                })

        return True

    # ============================================
    # SUBMISSION OPERATIONS
    # ============================================

    @staticmethod
    def create_submission(submission_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new submission"""
        submission_id = str(uuid.uuid4())
        now = datetime.now(timezone.utc).isoformat()

        submission = {
            'id': submission_id,
            'exam_id': submission_data['exam_id'],
            'user_id_or_session': submission_data.get('user_id_or_session', str(uuid.uuid4())),
            'started_at': now,
            'answers': submission_data.get('answers', {}),
            'progress_percent': 0,
            'last_playback_time': 0,
        }

        db.reference(f'submissions/{submission_id}').set(submission)

        # Update submission count
        exam = FirebaseService.get_exam(submission_data['exam_id'])
        if exam:
            FirebaseService.update_exam(submission_data['exam_id'], {
                'submission_count': exam.get('submission_count', 0) + 1
            })

        return submission

    @staticmethod
    def get_submission(submission_id: str) -> Optional[Dict[str, Any]]:
        """Get submission by ID"""
        return db.reference(f'submissions/{submission_id}').get()

    @staticmethod
    def update_submission(submission_id: str, submission_data: Dict[str, Any]) -> Dict[str, Any]:
        """Update submission"""
        db.reference(f'submissions/{submission_id}').update(submission_data)
        return FirebaseService.get_submission(submission_id)


    # ============================================
    # TRACK OPERATIONS (Migrated from MongoDB)
    # ============================================

    @staticmethod
    def create_track(track_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new track"""
        track_id = str(uuid.uuid4())
        now = datetime.now(timezone.utc).isoformat()

        track = {
            'id': track_id,
            'track_type': track_data.get('track_type', 'listening'),
            'title': track_data.get('title', 'Untitled Track'),
            'description': track_data.get('description', ''),
            'exam_id': track_data.get('exam_id'),
            'status': track_data.get('status', 'draft'),
            'created_at': now,
            'updated_at': now,
            'created_by': track_data.get('created_by', 'admin'),
            'metadata': track_data.get('metadata', {}),
            'tags': track_data.get('tags', []),
            'source': track_data.get('source', 'manual'),
        }

        db.reference(f'tracks/{track_id}').set(track)
        return track

    @staticmethod
    def get_track(track_id: str) -> Optional[Dict[str, Any]]:
        """Get track by ID"""
        return db.reference(f'tracks/{track_id}').get()

    @staticmethod
    def get_all_tracks(track_type: Optional[str] = None, status: Optional[str] = None) -> List[Dict[str, Any]]:
        """Get all tracks with optional filtering"""
        tracks_ref = db.reference('tracks')
        all_tracks = tracks_ref.get()

        if not all_tracks:
            return []

        tracks = list(all_tracks.values()) if isinstance(all_tracks, dict) else []

        # Filter by track_type if provided
        if track_type:
            tracks = [t for t in tracks if t.get('track_type') == track_type]

        # Filter by status if provided
        if status:
            tracks = [t for t in tracks if t.get('status') == status]

        # Sort by created_at descending
        tracks.sort(key=lambda x: x.get('created_at', ''), reverse=True)

        return tracks

    @staticmethod
    def update_track(track_id: str, track_data: Dict[str, Any]) -> Dict[str, Any]:
        """Update track"""
        track_data['updated_at'] = datetime.now(timezone.utc).isoformat()
        db.reference(f'tracks/{track_id}').update(track_data)
        return FirebaseService.get_track(track_id)

    @staticmethod
    def delete_track(track_id: str) -> bool:
        """Delete track"""
        db.reference(f'tracks/{track_id}').delete()
        return True

